core/src/streams.rs

```diff
@@ -161,4 +262,357 @@ impl<'gc> NetStream<'gc> {
+                    match FlvHeader::parse(&mut reader) {
+                        Ok(header) => {
+                            write.offset = reader.into_parts().1;
+                            write.preload_offset = write.offset;
+                            write.stream_type = Some(NetStreamType::Flv {
+                                header,
+                                stream: None,
+                                frame_id: 0,
+                            });
+                        }
+                        Err(FlvError::EndOfData) => return,
+                        Err(e) => {
+                            //TODO: Fire an error event to AS & stop playing too
+                            tracing::error!("FLV header parsing failed: {}", e);
+                            write.preload_offset = 3;
+                            return;
+                        }
+                    }
+                }
+                Some(_) => {
+                    //Unrecognized signature
+                    write.preload_offset = 3;
+                    return;
+                }
+                None => return, //Data not yet loaded
+            }
+        }
+
+        let end_time = write.stream_time + dt;
+        let mut end_of_video = false;
+
+        //At this point we should know our stream type.
+        if matches!(write.stream_type, Some(NetStreamType::Flv { .. })) {
+            let mut reader = FlvReader::from_parts(&write.buffer, write.offset);
+
+            loop {
+                let tag = FlvTag::parse(&mut reader);
+                if let Err(e) = tag {
+                    if matches!(e, FlvError::EndOfData) {
+                        //TODO: Check expected total length for streaming / progressive download
+                        end_of_video = true;
+                    } else {
+                        //Corrupt tag or out of data
+                        //TODO: Stop the stream so we don't repeatedly yield the same error
+                        //and fire an error event to AS
+                        tracing::error!("FLV tag parsing failed: {}", e);
+                    }
+
+                    break;
+                }
+
+                let tag = tag.expect("valid tag");
+                if tag.timestamp as f64 >= end_time {
+                    //All tags processed
+                    if let Err(e) = FlvTag::skip_back(&mut reader) {
+                        tracing::error!("FLV skip back failed: {}", e);
+                    }
+
+                    break;
+                }
+
+                let tag_needs_preloading = reader.stream_position().expect("valid position")
+                    as usize
+                    >= write.preload_offset;
+
+                match tag.data {
+                    FlvTagData::Audio(FlvAudioData { .. }) => {
+                        tracing::warn!("Stub: Stream audio processing");
+                    }
+                    FlvTagData::Video(FlvVideoData { codec_id, data, .. }) => {
```

Comment:
The function is somewhat long. Would it be helpful (or worse) to maybe put parts of this function into their own functions that is called from here? Like, a function that might be called `handle_flv_tag_data_video()` for this match?

But, one problem with that is how it would interfere with the `loop`, in case there are `return` or `break` or `continue` statements and the like. The usage of a `loop`, together with a lot of code, feels a bit difficult to handle to me, though I could be wrong, both on this point and on possible alternatives on organizing the code being sufficiently better to change to one of them.